### ISSUE TITLE
buildah bud : --creds with multi-stage build

### ISSUE THREAD

**Issue by gregoryboue** [containers/buildah#1329]
**Description**

When I execute a buildah bud with a multi-stage Dockerfile with : 

- A first FROM an image on public registry
- A second FROM an image on private registry
- With --creds for the private registry

I have an error on the first FROM execution : 

`Error determining manifest MIME type for docker://mypublicregistry/myimage:tag: unable to retrieve auth token: invalid username/password`

**Steps to reproduce the issue:**
1. Write this Dockerfile
```
FROM registry.docker.io/alpine as downloader
 
RUN echo test

FROM myprivateregistry/myimage:tag

RUN echo test

```

2. launch a buildah bud command with --creds option for the private registry

**Describe the results you received:**

`Error determining manifest MIME type for docker://mypublicregistry/myimage:tag: unable to retrieve auth token: invalid username/password`

**Describe the results you expected:**

No error on public registry authentication and authentication on the private registry

**Output of `rpm -q buildah` or `apt list buildah`:**

```
package buildah is not installed
```

**Output of `buildah version`:**

```
Version:         1.4
Go Version:      go1.9.4
Image Spec:      1.0.0
Runtime Spec:    1.0.0
CNI Spec:        0.4.0
libcni Version:  v0.7.0-alpha1
Git Commit:      608fa84
Built:           Mon Oct 22 14:53:00 2018
OS/Arch:         linux/amd64
```

**Output of `cat /etc/*release`:**

```
CentOS Linux release 7.3.1611 (Core) 
NAME="CentOS Linux"
VERSION="7 (Core)"
ID="centos"
ID_LIKE="rhel fedora"
VERSION_ID="7"
PRETTY_NAME="CentOS Linux 7 (Core)"
ANSI_COLOR="0;31"
CPE_NAME="cpe:/o:centos:centos:7"
HOME_URL="https://www.centos.org/"
BUG_REPORT_URL="https://bugs.centos.org/"

CENTOS_MANTISBT_PROJECT="CentOS-7"
CENTOS_MANTISBT_PROJECT_VERSION="7"
REDHAT_SUPPORT_PRODUCT="centos"
REDHAT_SUPPORT_PRODUCT_VERSION="7"

CentOS Linux release 7.3.1611 (Core) 
CentOS Linux release 7.3.1611 (Core)
```

**Output of `uname -a`:**

```
Linux jenkins-slave-52fwh-mhkn5 3.10.0-862.9.1.el7.x86_64 #1 SMP Wed Jun 27 04:30:39 EDT 2018 x86_64 x86_64 x86_64 GNU/Linux
```

**Output of `cat /etc/containers/storage.conf`:**

```
# storage.conf is the configuration file for all tools
# that share the containers/storage libraries
# See man 5 containers-storage.conf for more information
# The "container storage" table contains all of the server options.
[storage]

# Default Storage Driver
driver = "overlay"

# Temporary storage location
runroot = "/var/run/containers/storage"

# Primary Read/Write location of container storage
graphroot = "/var/lib/containers/storage"

[storage.options]
# Storage options to be passed to underlying storage drivers

# AdditionalImageStores is used to pass paths to additional Read/Only image stores
# Must be comma separated list.
additionalimagestores = [
]

# Size is used to set a maximum size of the container image.  Only supported by
# certain container storage drivers.
size = ""

# OverrideKernelCheck tells the driver to ignore kernel checks based on kernel version
override_kernel_check = "true"

# Remap-UIDs/GIDs is the mapping from UIDs/GIDs as they should appear inside of
# a container, to UIDs/GIDs as they should appear outside of the container, and
# the length of the range of UIDs/GIDs.  Additional mapped sets can be listed
# and will be heeded by libraries, but there are limits to the number of
# mappings which the kernel will allow when you later attempt to run a
# container.
#
# remap-uids = 0:1668442479:65536
# remap-gids = 0:1668442479:65536

# Remap-User/Group is a name which can be used to look up one or more UID/GID
# ranges in the /etc/subuid or /etc/subgid file.  Mappings are set up starting
# with an in-container ID of 0 and the a host-level ID taken from the lowest
# range that matches the specified name, and using the length of that range.
# Additional ranges are then assigned, using the ranges which specify the
# lowest host-level IDs first, to the lowest not-yet-mapped container-level ID,
# until all of the entries have been used for maps.
#
# remap-user = "storage"
# remap-group = "storage"

[storage.options.thinpool]
# Storage Options for thinpool

# autoextend_percent determines the amount by which pool needs to be
# grown. This is specified in terms of % of pool size. So a value of 20 means
# that when threshold is hit, pool will be grown by 20% of existing
# pool size.
# autoextend_percent = "20"

# autoextend_threshold determines the pool extension threshold in terms
# of percentage of pool size. For example, if threshold is 60, that means when
# pool is 60% full, threshold has been hit.
# autoextend_threshold = "80"

# basesize specifies the size to use when creating the base device, which
# limits the size of images and containers.
# basesize = "10G"

# blocksize specifies a custom blocksize to use for the thin pool.
# blocksize="64k"

# directlvm_device specifies a custom block storage device to use for the
# thin pool. Required if you setup devicemapper
# directlvm_device = ""

# directlvm_device_force wipes device even if device already has a filesystem
# directlvm_device_force = "True"

# fs specifies the filesystem type to use for the base device.
# fs="xfs"

# log_level sets the log level of devicemapper.
# 0: LogLevelSuppress 0 (Default)
# 2: LogLevelFatal
# 3: LogLevelErr
# 4: LogLevelWarn
# 5: LogLevelNotice
# 6: LogLevelInfo
# 7: LogLevelDebug
# log_level = "7"

# min_free_space specifies the min free space percent in a thin pool require for
# new device creation to succeed. Valid values are from 0% - 99%.
# Value 0% disables
# min_free_space = "10%"

# mkfsarg specifies extra mkfs arguments to be used when creating the base
# device.
# mkfsarg = ""

# mountopt specifies extra mount options used when mounting the thin devices.
# mountopt = ""

# use_deferred_removal Marking device for deferred removal
# use_deferred_removal = "True"

# use_deferred_deletion Marking device for deferred deletion
# use_deferred_deletion = "True"

# xfs_nospace_max_retries specifies the maximum number of retries XFS should
# attempt to complete IO when ENOSPC (no space) error is returned by
# underlying storage device.
# xfs_nospace_max_retries = "0"
```

**Comment by TomSweeneyRedHat** [comment]
Hi @gregoryboue , thanks for the issue.  

Two quick thoughts.  First is if you can, please upgrade to the latest Buildah.  We're currently at v1.6 and upstream in GitHub is v1.7.  We'd a number of changes that went in for v1.5 surrounding credential fixes that might address this issue and a few more in 1.6.

Secondly rather than using the --creds option with the buildh, can you install podman and try 'podman login' into your private registry and retry the build?  Or if by chance you still have Docker on the machine, try 'docker login' instead?

**Comment by gregoryboue** [comment]
Hi @TomSweeneyRedHat , thanks for the answer, I did another test with this buildah version : 

```
11:19:48 + buildah version
11:19:48 Version:         1.6
11:19:48 Go Version:      go1.10.4
11:19:48 Image Spec:      1.0.0
11:19:48 Runtime Spec:    1.0.0
11:19:48 CNI Spec:        0.4.0
11:19:48 libcni Version:  v0.7.0-alpha1
11:19:48 Git Commit:      5f95bd9
11:19:48 Built:           Tue Feb 12 12:47:11 2019
11:19:48 OS/Arch:         linux/amd64
```

The result is the same, for the second way to do it with podman login, it can be a solution but in a CI build context is not so easy to managed it

**Comment by rhatdan** [comment]
So we need buildah login, to make this work.

**Comment by QiWang19** [comment]
--cred username:password works for me with current version
```
$ buildah version
Version:         1.8-dev
Go Version:      go1.11.2
Image Spec:      1.0.0
Runtime Spec:    1.0.0
CNI Spec:        0.4.0
libcni Version:  v0.7.0-alpha1
Git Commit:      f83e69c3
Built:           Wed Mar 20 10:41:56 2019
OS/Arch:         linux/amd64
```

**Comment by TomSweeneyRedHat** [comment]
@QiWang19 did you test using a private registry and doing the multistage build as noted in the description?  My guess is the value from --cred wasn't getting to the second FROM in the Dockerfile for some reason.  I'd be happy if that's not the case now, but I'm a bit suspicious.

**Comment by QiWang19** [comment]
It's a bug

**Comment by QiWang19** [comment]
I copied `podman login` code to buildah, why podman login and buildah login write cresdential to /run/containers/0/auth.json, but buildah pull reads from /var/run/containers/storage/containers/auth.json? what might be the reason?

**Comment by TomSweeneyRedHat** [comment]
@QiWang19 I'm not sure what you mean by copying the podman login code to buildah, but referencing `/var/run/containers/storage/containers/auth.json` instead of `/run/containers/0/auth.json` sounds like you might have found a bug.  @umohnani8 any thoughts?  I think it should all be looking at `/run/containers/{uid}/containers` (${XDG_RUNTIME_DIR}/containers).

**Comment by QiWang19** [comment]
@TomSweeneyRedHat I just mean I followed the command above (need `buildah login`),  copied the code of podman login to make `buildah login` command work. ^_^

**Comment by umohnani8** [comment]
@QiWang19 @TomSweeneyRedHat not sure. It used to look at `/var/run/containers...` until we decided to move over to `/run/containers/uid`. Maybe some remnants of the old code?

**Comment by gregoryboue** [comment]
Sorry for the lack of response.
In my case the context is a CI/CD environment to build and push images.
The build environment is created for each build.

So the solution of the podman /buildah login is not a solution in my case.

I have a Docker Trusted Registry in my company.

When i perform a multistage build with the first FROM which pull a public repository image and the second FROM which pull a private repository image then the error occured on the first stage.

**Comment by QiWang19** [comment]
@gregoryboue  is it helpful to use `--password-stdin` to pass the password from Stdin?

**Comment by gregoryboue** [comment]
@QiWang19 , in our case it is not helpful, I did a mistake in my last comment, the registries were different.
In our case we will impose to use the same registry for all the FROM command and it will be ok

**Comment by QiWang19** [comment]
Does it help to use buildah login to the different registries before starting the build?
Do I need to reopen this issue @TomSweeneyRedHat  @rhatdan  ?
